utils/iterator/deduplicate.go

```diff
@@ -0,0 +1,40 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package iterator
+
+import "github.com/ava-labs/avalanchego/utils/set"
+
+var _ Iterator[any] = (*deduplicator[any])(nil)
+
+type deduplicator[T comparable] struct {
```

Comment:
```
type deduplicator[T comparable] struct {
	Iterator[T]
	seen set.Set[T]
}
```

Wondering if it makes sense to just embed the iterator and drop Value and Release

Reply:
I works locally for me: 

```
index b36c1e0a1..4ba769c12 100644
--- a/utils/iterator/deduplicate.go
+++ b/utils/iterator/deduplicate.go
@@ -8,7 +8,7 @@ import "github.com/ava-labs/avalanchego/utils/set"
 var _ Iterator[any] = (*deduplicator[any])(nil)
 
 type deduplicator[T comparable] struct {
-       it   Iterator[T]
+       Iterator[T]
        seen set.Set[T]
 }
 
@@ -16,13 +16,13 @@ type deduplicator[T comparable] struct {
 // been returned from [it].
 func Deduplicate[T comparable](it Iterator[T]) Iterator[T] {
        return &deduplicator[T]{
-               it: it,
+               Iterator: it,
        }
 }
 
 func (i *deduplicator[_]) Next() bool {
-       for i.it.Next() {
-               element := i.it.Value()
+       for i.Iterator.Next() {
+               element := i.Iterator.Value()
                if !i.seen.Contains(element) {
                        i.seen.Add(element)
                        return true
@@ -30,11 +30,3 @@ func (i *deduplicator[_]) Next() bool {
        }
        return false
 }
-
-func (i *deduplicator[T]) Value() T {
-       return i.it.Value()
-}
-
-func (i *deduplicator[_]) Release() {
-       i.it.Release()
-}
```

Reply:
I removed this custom iterator implementation and reused `Filter`.

Reply:
Embedding does work - but it has introduced subtle bugs in the past (because changing the signature of `Next` to `NewNext` or something wouldn't cause the code to fail to compile, but would cause the implementation to be wrong)...

This is why we use `Unsafe.*Server` rather than `Unimplemented.*Server` in our grpc services.

Reply:
Reminds me of `InsecureSkipVerify`